File: Queen'sAttack.py
# 1) take an array and construct it like chessboard and 
# place queen on chess board

import numpy as np

# co ordinates of queens
r_q = 4
c_q = 3


# n of chess board
n = 5

r_q = n -r_q
c_q = c_q - 1

array = np.array([[0,0,0,0,0], [0,0,0,0,0],[0,0,0,0,0],[0,0,0,0,0], [0,0,0,0,0]])

# placing queen on the chess board
array[r_q][c_q] = 1

# 2) placing obstacles on chessboard
k = 3

obstacles_rows = 3
obstacles_columns = 2

obstacles = np.array([[5,5],[4,2],[2,3]])

for i in range(obstacles_rows):

    obstacles[i][0] = n - obstacles[i][0]
    obstacles[i][1] = obstacles[i][1] - 1

# placing obstacles on chessboard
ro = 0
co = 0

for i in range(obstacles_rows):

    ro = obstacles[i][0]
    co = obstacles[i][1]

    array[ro][co] = -1

#   2.1) traversing the downwards [done]

for i in range(r_q,n):

   if(array[i][c_q] != -1):
       array[i][c_q] = 1
   else:
       break   

#   2.2) traversing upwards [done]
for i in range(r_q, -1, -1):

   if(array[i][c_q] != -1):
       array[i][c_q] = 1
   else:
       break

#   2.3) traversing right side [done]
for i in range(c_q, n):

   if(array[r_q][i] != -1):
       array[r_q][i] = 1
   else:
       break

#   2.4) traversing left side pdone]
for i in range(c_q, -1, -1):

   if(array[r_q][i] != -1):
       array[r_q][i] = 1
   else:
       break

#   2.5) traversing diagonal downwards towards right
i = r_q
j = c_q

while (i < n and j < n):

    if(array[i][j] != -1):
    
        array[i][j] = 1
        i += 1
        j += 1
    else:
        break

#   2.6) traversing diagonal downwards towards left
i =r_q
j =c_q

while (i < n and j >= 0):

    if(array[i][j] != -1):
        array[i][j] = 1
        i += 1
        j -= 1
    else:

        break


#  2.7) traversing diagonal upwards towards left
i =r_q
j =c_q

while (i >= 0 and j >= 0):

    if(array[i][j] != -1):

        array[i][j] = 1
        i -= 1
        j -= 1
    
    else:
        break

#  2.8) traversing diagonal upwards towards right
i =r_q
j =c_q

while (i >= 0 and j < n):

    if(array[i][j] != -1):
        array[i][j] = 1
        i -= 1
        j += 1
    
    else:
        break

   
# 4) traversing through array and finding out how many squares did queen traverse

count = 0

for i in range(n):

    for i in range(n):
    
        if(array[i][j] == 1):
            count += 1

print (count)

File: Queen'sATtack.c
#include <stdio.h>
#include <stdlib.h>


void traversing(int r_q, int c_q, int **chessboard, int n)
{

    // 1) traversing queen upward
    int row = r_q - 1;
    int col = c_q;

    while((row < n && row >= 0))
    {

        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            row -= 1;
        }
        else
        {
            break;
        }
    }
        

    // 2) traversing queen downward
    row = r_q+1;
    col = c_q;


    while((row < n && row >= 0))
    {
        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            row += 1;
        }
        else
        {
            break;
        }
    }

    
    // 3) traversing queen righside
    row = r_q;
    col = c_q+1;

    
    while((col < n && col >= 0))
    {

        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            col += 1;

        }
        else
        {
            break;
        }
    
    }

    // 4) traversing queen leftside
    row = r_q;
    col = c_q-1;

    while((col < n && col >= 0))
    {
        
        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            col -= 1;
        }
        else
        {
            break;
        }

    }
    
    // 5) traversing queen upward right
    row = r_q-1;
    col = c_q+1;
    
    while((col < n && col >= 0) && (row < n && row >= 0))
    {
        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            col += 1;
            row -= 1;
        }
        else
        {
            break;
        }
    }
        
        
    // 6) traversing queen downward right
    row = r_q+1;
    col = c_q+1;
    
    while((col < n && col >= 0) && (row < n && row >= 0))
    {
        
        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            col += 1;
            row += 1;
        }
        else
        {
            break;
        }
    }

    // 7) traversing queen upward left
    row = r_q-1;
    col = c_q-1;
    
    while((col < n && col >= 0) && (row < n && row >= 0))
    {
        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            col -= 1;
            row -= 1;
        }
        else
        {
            break;
        }
    }
        
        
    // 8) traversing queen downward left
    row = r_q+1;
    col = c_q-1;
    
    while((col < n && col >= 0) && (row < n && row >= 0))
    {

        
        if(chessboard[row][col] != -1)
        {
            chessboard[row][col] = 1;
            col -= 1;
            row += 1;
        }
        else
        {
            break;
        }
    }
}

int counting(int **chessboard, int n)
{
    int count = 0;

    for(int x= 0; x<n; x++)
    {
        for(int y= 0; y < n; y ++)
        {
            if(chessboard[x][y] == 1)
            {
                count += 1;
            }
        }
    }

    return count;
}


int queensAttack(int n, int k, int r_q, int c_q, int obstacles_rows, int obstacles_columns, int** obstacles)
{


    // 1) create a chessboard with linked_list
    int *chessboard[n];
    for(int i = 0;i < n; i++)
    {
        chessboard[i] = (int*) calloc(n, sizeof(int)); 
    }


    // 2) change the co ordinates of the obstacles to the chessboard we've created
    for(int i = 0; i< k; i ++)
    {
        // 2.1) subtract row of obstacle from n to change the co ordinates of the obstacle
        obstacles[i][0] = n - obstacles[i][0];

        // 2.2) subtract 1 from the row to change the co ordinates of the obstacle
        obstacles[i][1] = obstacles[i][1] - 1;
    }

    // 3) place obstacles on the chessboard
    for(int y = 0; y < k; y ++)
    {
        chessboard[obstacles[y][0]][obstacles[y][1]] = -1;
    }


    // 4) change the co ordinates of the queen to the chessboard
    
    // 4.1) subtract row of queen from n to change the  co ordinates of the queen
    r_q = n - r_q;

    // 4.2) subtract 1 from the row to change the co ordinates of the queen
    c_q = c_q - 1;

    // 5) place the queen on the chessboard
    chessboard[r_q][c_q] = 2;

    // 6) traverse the queen in 8 directions
    traversing(r_q, c_q, chessboard, n);

    // 7) find the no of blocks queen has traversed
    return counting(chessboard, n);
}

int main()
{
    int obstacles[3][2] = {{5,5},{4,2},{2,3}};
    printf("%d",queensAttack(5,3,4,3,3,2,obstacles));
    
}


